DSharpPlus/Net/RestClient.cs

```diff
@@ -14,27 +15,35 @@
 
 namespace DSharpPlus.Net
 {
+    /// <summary>
+    /// Creates an instance of a Http client implementation.
+    /// </summary>
+    /// <param name="messageHandler">Message handler to use for the new Http client instance.</param>
+    /// <returns>Constructed Http client implementation.</returns>
+    public delegate HttpClient HttpClientFactoryDelegate(HttpMessageHandler messageHandler);
+
     /// <summary>
     /// Represents a client used to make REST requests.
     /// </summary>
     internal sealed class RestClient
     {
         private static UTF8Encoding UTF8 { get; } = new UTF8Encoding(false);
         private static Regex RouteArgumentRegex { get; } = new Regex(@":([a-z_]+)");
+        private static TimeSpan RateLimitResetBuffer { get; } = TimeSpan.FromMilliseconds(10);
 
         private BaseDiscordClient Discord { get; }
         private HttpClient HttpClient { get; }
-        private HashSet<RateLimitBucket> Buckets { get; }
-        private SemaphoreSlim RequestSemaphore { get; }
+        private ConcurrentDictionary<string, RateLimitBucket> Buckets { get; }
+        private ManualResetEventSlim GlobalRateLimitEvent { get; }
```

Comment:
If you're working in async code, you go async all the way. If you use synchronous waits, you will cause tasks to stall. Use a `TaskCompletionSource<T>` or another form of asynchronous waiting here.

Reply:
The .met ManualResetEvent(Slim) has no async Wait method this would require using a library like AsyncEx or a custom implementation of such a event but since this only blocks when the global ratelimit has been hit i refrained from taking either action and went with the synchronous implementation included in the framework.

Reply:
Yes, but you can perform an async wait by using a `TaskCompletionSource<T>`, specifically awaiting its `.Task` property, and using `.SetResult(T)` to set the task to completed state.

Reply:
Should be fixed, I added a AsyncManualResetEvent (I hope in the right place) for this